src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime, boolean inclusive) {
+		this.uEnd = null;
+		this.uEndTime = endTime;
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the maximum number of messages to get. Defaults to -1 (no limit).
+	 *
+	 * @param count max number of messages to get
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder withMaxCount(int count) {
+		if (count < 1) throw new IllegalArgumentException();
+		this.count = count;
+		return this;
+	}
+
+	/**
+	 * Returns a stream of the message history.
+	 *
+	 * @return the stream of the message history
+	 * @see MessageHistoryIterator
+	 */
+	public Stream<IMessage> stream() {
+
+		buildRange();
+
+		return new MessageHistoryIterator(range).stream();
+	}
+
+	/**
+	 * Returns an iterator for this message history.
+	 * NOTE: This object can be put directly in a for-each loop.
+	 *
+	 * @return an Iterator instance for the message history
+	 */
+	@Override
+	public Iterator<IMessage> iterator() {
+
+		buildRange();
+
+		return new MessageHistoryIterator(range);
+	}
+
+	/**
+	 * Requests a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will not block the thread during execution. Use {@link MessageHistoryBuilder#get()} to block the
+	 * thread.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public RequestBuffer.RequestFuture<MessageHistory> request() {
+		return RequestBuffer.request(() -> {
+			return this.get();
+		});
+	}
+
+	/**
+	 * Builds a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will block the thread until all messages are received. Use {@link MessageHistoryBuilder#request()}
+	 * to request this message history without blocking.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public MessageHistory get() {
+
+		// Get cache from the channel, sorted
+		cached = channel.messages.stream().sorted(MessageComparator.REVERSED).collect(Collectors.toList());
+
+		buildRange();
+
+		// Initialize messages
+		messages = count < 0 ?
+				new ArrayList<>(MESSAGE_CHUNK_COUNT) :
+				new ArrayList<>(count);
+
+		// Checks for special case for efficiency- if a range is reverse ordered (start earlier in history than end) but
+		// count unlimited, then it is equivalent to switching endpoints and reversing the collection afterwards
+		if (range.isChronological() && count < 0) {
+
+			range = new MessageHistoryRange(channel,
+					uEnd == null ?
+							new MessageHistoryRange.Endpoint(uEndTime, uIncludeEnd) :
+							new MessageHistoryRange.Endpoint(uEnd, uIncludeEnd),
+					uStart == null ?
+							new MessageHistoryRange.Endpoint(uStartTime, uIncludeStart) :
+							new MessageHistoryRange.Endpoint(uStart, uIncludeStart)
+			);
+
+			fetch();
+
+			Collections.reverse(messages);
+		} else {
+			fetch();
+		}
+
+		return new MessageHistory(messages);
+	}
+
+	private void buildRange() {
```

Comment:
~~I think it's pretty readable.~~ I can replace with nested if's, if you'd prefer.